### PR TITLE
wasm2js: Support for exported memory

### PR DESCRIPTION
**Pull request by sbc100**
This change is required for https://github.com/emscripten-core/emscripten/pull/12323

### CODE REVIEW

**Comment by sbc100** [comment]
I think this good to land now.

File: src/wasm2js.h [hunk 1]
```diff
@@ -2445,6 +2459,18 @@ void Wasm2JSGlue::emitPreES6() {
 }
 
 void Wasm2JSGlue::emitPost() {
+  // Create a helper bufferView to access the buffer if we need one. We use it
+  // for creating memory segments if we have any (we may not if the segments are
+  // shipped in a side .mem file, for example), and also in bulk memory
+  // operations.
+  // This will get assigned during `asmFunc` (and potentailly re-assiged
```
Comment (kripken) [inline src/wasm2js.h:2466]:
```suggestion
  // This will get assigned during `asmFunc` (and potentially re-assigned
```

File: src/wasm2js.h [hunk 2]
```diff
@@ -2445,6 +2459,18 @@ void Wasm2JSGlue::emitPreES6() {
...
+  // during __wasm_memory_grow
```
Comment (kripken) [inline src/wasm2js.h:2467]:
```suggestion
  // during __wasm_memory_grow).
```

File: src/wasm2js.h [hunk 3]
```diff
@@ -503,6 +512,10 @@ void Wasm2JSBuilder::addBasics(Ref ast, Module* wasm) {
     addHeap(HEAPU32, UINT32ARRAY);
     addHeap(HEAPF32, FLOAT32ARRAY);
     addHeap(HEAPF64, FLOAT64ARRAY);
+    ast->push_back(
+      ValueBuilder::makeBinary(ValueBuilder::makeName("bufferView"),
+                               SET,
+                               ValueBuilder::makeName(HEAPU8)));
```
Comment (kripken) [inline src/wasm2js.h:518]: this creation of `bufferView` looks unconditional - whenever there is a memory at all. why is another one needed elsewhere?

File: test/wasm2js/atomic_fence.2asm.js [hunk 4]
```diff
@@ -55,10 +57,7 @@ function asmFunc(env, buffer) {
  };
 }
 
-var memasmFunc = new ArrayBuffer(1507328);
-var bufferView = new Uint8Array(memasmFunc);
+var bufferView;
```
Comment (kripken) [inline test/wasm2js/atomic_fence.2asm.js:60]: can we not remove this `bufferView` entirely, and just use `HEAP8` which is guaranteed to exist?

Comment (sbc100) [reply test/wasm2js/atomic_fence.2asm.js:60]: See my `TODO: We should probably just share a single HEAPU8 var.` above. 

I was thinking that could be a followup.     Keeping this change slightly more limited.

Comment (kripken) [reply test/wasm2js/atomic_fence.2asm.js:60]: This does make it harder to review, though, see the comment above with `unconditional` - it looks like the current PR is not emitting `bufferView` properly, but I may be missing something?

Comment (sbc100) [reply test/wasm2js/atomic_fence.2asm.js:60]: If we start to share HEAPU8 with the containing scope is that going to interfere with the emscripten-side that also defines a HEAPU8 itself?   I mean that should always point to the same thing anyway so maybe its fine?

File: src/wasm2js.h (hunk 3 above)
Comment (sbc100) [reply src/wasm2js.h:518]: This is not a creation of anything but an assignment to the outer scoped `bufferView`

But you are right I think it needs to be conditional in the same way that bufferView itself is conditionally declared in the outer scope.

File: test/wasm2js/atomic_fence.2asm.js (hunk 4 above)
Comment (sbc100) [reply test/wasm2js/atomic_fence.2asm.js:60]: Yes it looks like the wasm2js output shares a scope with the rest of emscripten, but `asmFunc` currently creates it own function-local versions of HEAP8, etc.   This means that in emscripten output there are two copies of all these views today.   

We should possibly merge those two and (i.e. have the output HEAP8 be the same as the inner HEAP8), but I'd rather not worry about that as part of this change.

File: src/wasm2js.h (hunk 3 above)
Comment (kripken) [reply src/wasm2js.h]: Oh, I see, I was misreading it, sorry.

**Review by kripken — APPROVED** [review]
This is fairly confusing (not because of you, but because the wasm2js output is complex with the different modes) but reading the outputs I've convinced myself it is correct.

**Pull request merged**